pkg/pool-hooks/contracts/StableSurgeHookExample.sol

```diff
@@ -0,0 +1,243 @@
+ * @dev Uses the dynamic fee mechanism to apply a directional fee.
+ */
+contract StableSurgeHookExample is BaseHooks, VaultGuard {
+    using FixedPoint for uint256;
+
+    uint256 public constant MAX_SURGE_FEE_PERCENTAGE = 95e16; // 95%
+
+    // The default threshold, above which surging will occur.
+    uint256 private immutable _defaultSurgeThresholdPercentage;
+
+    // Store the current threshold for each pool.
+    mapping(address pool => uint256 threshold) private _surgeThresholdPercentage;
+
+    /**
+     * @notice A new `StableSurgeHookExample` contract has been registered successfully.
+     * @dev If the registration fails the call will revert, so there will be no event.
+     * @param pool The pool on which the hook was registered
+     */
+    event StableSurgeHookExampleRegistered(address indexed pool);
+
+    /**
+     * @notice The threshold percentage has been changed for a pool in a `StableSurgeHookExample` contract.
+     * @dev Note, the initial threshold percentage is set on deployment and an event is emitted.
+     * @param pool The pool for which the threshold percentage has been changed
+     * @param newSurgeThresholdPercentage The new threshold percentage
+     */
+    event ThresholdSurgePercentageChanged(address indexed pool, uint256 newSurgeThresholdPercentage);
+
+    /// @notice The sender does not have permission to call a function.
+    error SenderNotAllowed();
+
+    /// @notice The threshold must be a valid percentage value.
+    error InvalidSurgeThresholdPercentage();
+
+    constructor(IVault vault, uint256 defaultSurgeThresholdPercentage) VaultGuard(vault) {
+        _ensureValidPercentage(defaultSurgeThresholdPercentage);
+
+        _defaultSurgeThresholdPercentage = defaultSurgeThresholdPercentage;
+    }
+
+    /// @inheritdoc IHooks
+    function getHookFlags() public pure override returns (HookFlags memory hookFlags) {
+        hookFlags.shouldCallComputeDynamicSwapFee = true;
+    }
+
+    /**
+     * @notice Getter for the default surge threshold percentage.
+     * @return surgeThresholdPercentage The default surge threshold percentage for this hook contract
+     */
+    function getDefaultSurgeThresholdPercentage() external view returns (uint256) {
+        return _defaultSurgeThresholdPercentage;
+    }
+
+    /**
+     * @notice Getter for the surge threshold percentage for a pool.
+     * @param pool The pool for which the surge threshold percentage is requested
+     * @return surgeThresholdPercentage The surge threshold percentage for the pool
+     */
+    function getSurgeThresholdPercentage(address pool) external view returns (uint256) {
+        return _surgeThresholdPercentage[pool];
+    }
+
+    /// @inheritdoc IHooks
+    function onRegister(
+        address,
+        address pool,
+        TokenConfig[] memory,
+        LiquidityManagement calldata
+    ) public override onlyVault returns (bool) {
+        emit StableSurgeHookExampleRegistered(pool);
```

Comment:
Check for StablePoolFactory?

Reply:
I don't think we should limit factories. I mean custom factories should be able to use this hook. WDYT about it?

Reply:
My understanding is this is an official Balancer hook that will be released in production. If that's the case, the straightforward thing would be to create a new StablePoolWithSurgeFactory that deploys this hook on deployment, then passes the hook address and its own address as the "allowedFactory" when creating new StablePools.

People wanting to use it with their own custom pools would do the same in their custom pool factory.

It really only works with StablePools. If we wanted to make it more generic, we could deploy it independently, and have an internal `mapping(address factory -> bool allowed)` with a permissioned call to allow governance to maintain a set of allowed factories. We'd add our StablePoolFactory first, and then others if they became available and were accepted by the community.

Reply:
What's the harm of not checking that the pool using this hook is not from Balancer's stable pool factory? And, is it really necessary to make a permissioned call to register other allowed factories?

This hook only uses dynamic swap fee hook, it does not move tokens around, would there be a possible exploit in using this hook in a place that it shouldn't be used? We could probably document that this hook is supposed to be used in a StablePool (or weighted pools with the same weight, it'd also work), but overloading the governance with a permissioned call to allow factories in a hook that, at first sight, is not harmful, seems to be overkill. (I'm probably oversimplifying and missing some very important detail, I just want to know what detail it is)

Reply:
I'd say it's not dangerous, but in `onComputeDynamicSwapFeePercentage` we're treating the pool as a stable pool and calling functions that are not present elsewhere.

If the factory is not the right one, it'll always revert. That can be prevented if we block registrations by mistake here. I would add the check.

Reply:
It seems more reasonable to do this after https://github.com/balancer/balancer-v3-monorepo/pulls